File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = float('-inf')
        st = []
        
        for i, h in enumerate(heights):
            start = i
            
            while st and st[-1][1] > h:
                maxArea = max(maxArea, (st[-1][1] * (i - st[-1][0])))
                start = st[-1][0]
                st.pop()

            st.append([start, h])
        
        n = len(heights)
        while st:
            maxArea = max(maxArea, (st[-1][1] * (n - st[-1][0])))
            st.pop()
        return maxArea

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java

class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Pair<Integer, Integer>> st = new Stack<>();
        int maxArea = 0;
        int n = heights.length;
        
        for (int i = 0; i < n; i++) {
            int start = i;
            while (!st.isEmpty() && st.peek().getValue() > heights[i]) {
                Pair<Integer, Integer> pair = st.pop();
                int height = pair.getValue();
                int width = i - pair.getKey();
                maxArea = Math.max(maxArea, height * width);
                start = pair.getKey();
            }
            st.push(new Pair<>(start, heights[i]));
        }
        
        // Process any remaining elements in the stack
        while (!st.isEmpty()) {
            Pair<Integer, Integer> pair = st.pop();
            int height = pair.getValue();
            int width = n - pair.getKey();
            maxArea = Math.max(maxArea, height * width);
        }
        
        return maxArea;
    }
}
